src/python/pants/backend/kotlin/compile/kotlinc_test.py

```diff
@@ -0,0 +1,583 @@
+    }
+    """
+)
+
+
+KOTLIN_LIB_MAIN_SOURCE = dedent(
+    """
+    package org.pantsbuild.example
+
+    import org.pantsbuild.example.lib.C
+
+    fun main(args: Array<String>) {
+        val c = C()
+        println(c.hello)
+    }
+    """
+)
+
+
+@maybe_skip_jdk_test
+def test_compile_no_deps(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'lib',
+                )
+                """
+            ),
+            "3rdparty/jvm/BUILD": DEFAULT_KOTLIN_STDLIB_TARGET,
+            "3rdparty/jvm/default.lock": DEFAULT_LOCKFILE,
+            "ExampleLib.kt": KOTLIN_LIB_SOURCE,
+        }
+    )
+    coarsened_target = expect_single_expanded_coarsened_target(
+        rule_runner, Address(spec_path="", target_name="lib")
+    )
+
+    classpath = rule_runner.request(
+        RenderedClasspath,
+        [CompileKotlinSourceRequest(component=coarsened_target, resolve=make_resolve(rule_runner))],
+    )
+    assert classpath.content == {
+        ".ExampleLib.kt.lib.kotlin.jar": {
+            "META-INF/MANIFEST.MF",
+            "META-INF/main.kotlin_module",
+            "org/pantsbuild/example/lib/C.class",
+        }
+    }
+
+    # Additionally validate that `check` works.
+    check_results = rule_runner.request(
+        CheckResults,
+        [
+            KotlincCheckRequest(
+                [KotlincCheckRequest.field_set_type.create(coarsened_target.representative)]
+            )
+        ],
+    )
+    assert len(check_results.results) == 1
+    check_result = check_results.results[0]
+    assert check_result.exit_code == 0
+
+
+@maybe_skip_jdk_test
+def test_compile_no_deps_jdk_12(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'lib',
+                    jdk = 'adopt:1.12',
+                )
+                """
+            ),
+            "3rdparty/jvm/BUILD": DEFAULT_KOTLIN_STDLIB_TARGET,
+            "3rdparty/jvm/default.lock": DEFAULT_LOCKFILE,
+            "ExampleLib.scala": KOTLIN_LIB_JDK12_SOURCE,
+        }
+    )
+    coarsened_target = expect_single_expanded_coarsened_target(
+        rule_runner, Address(spec_path="", target_name="lib")
+    )
+
+    rule_runner.request(
+        RenderedClasspath,
+        [CompileKotlinSourceRequest(component=coarsened_target, resolve=make_resolve(rule_runner))],
+    )
+
+
+@maybe_skip_jdk_test
+def test_compile_jdk_12_file_fails_on_jdk_11(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                kotlin_sources(
+                    name = 'lib',
+                    jdk = 'adopt:1.11',
+                )
+                """
+            ),
+            "3rdparty/jvm/BUILD": DEFAULT_KOTLIN_STDLIB_TARGET,
```

Comment:
In future: I find it easier to start with fixtures like this as methods, as it makes it easier to template them and to add additional default arguments that affect their behavior without updating callsites.

Reply:
Except I started with copy-pasta from the Scala rules. :)